### PR TITLE
feat: add segment writer client

### PR DESCRIPTION
**Pull request by kolesnikovae**
Depends on https://github.com/grafana/pyroscope/pull/3506
Closes: [#199](https://github.com/grafana/pyroscope-squad/issues/199), [#185](https://github.com/grafana/pyroscope-squad/issues/185) (private)

The change introduces the segment writer client and the write-path router components. A basic distribution algorithm based on jump consistent hashing was also implemented: dynamic sizing is to be implemented later.

Note that this also moves pprof-split out of the distributor – to be implemented in the segment writer in a follow-up PR.

### CODE REVIEW

**Review by korniltsev — APPROVED** [review]

**Comment by kolesnikovae** [comment]
Depends on #3506

File: pkg/util/circuitbreaker/cirquit_breaker.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+func UnaryClientInterceptor(cb *gobreaker.CircuitBreaker[any]) grpc.UnaryClientInterceptor {
+	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		_, err := cb.Execute(func() (interface{}, error) {
+			return nil, invoker(ctx, method, req, reply, cc, opts...)
+		})
+		// gobreaker returns unwrapped errors.
+		switch err {
+		case nil:
+		case gobreaker.ErrOpenState,
```
Comment (korniltsev) [inline pkg/util/circuitbreaker/cirquit_breaker.go:21]: nit: should this be checked with error.Is  / errors.As or is it guaranteed the library does not return wrapped errors?

**Review by korniltsev — APPROVED** [review]
I love have the routing is separated from the distributor

**Comment by korniltsev** [comment]
and sorry for merge conflicts :shrug:

File: pkg/util/circuitbreaker/cirquit_breaker.go (hunk 1 above)
Comment (kolesnikovae) [reply pkg/util/circuitbreaker/cirquit_breaker.go:21]: Thanks for asking! It is guaranteed that the errors are not wrapped. I was on the fence but finally settled on this. I think if there are questions, it's better to go another way

Comment (kolesnikovae) [reply pkg/util/circuitbreaker/cirquit_breaker.go]: Addressed in [clarify gobreaker error handling](https://github.com/grafana/pyroscope/pull/3505/commits/c31c9c8b679291a0a663f947f8fd1d9054c384f9)

**Pull request merged**